crates/ark/src/lsp/snapshots/indent.R

```diff
@@ -1098,11 +1098,11 @@ fun_call(object1 + object2 ~ object3 +
            argument)
 
 ## 32
-fun_call(object ~
+fun_call(~ object
   )
 
 ## 33
-fun_call(object +
+fun_call(object + object2
   )
```

Comment:
I'm not sure about all these changes to `indent.R`. Here these tests were intentionally checking invalid code. An indenter should ideally work no matter the validity of the code (i.e. partially written code as in these examples).

I guess it's fine if you want to take it as is for formatter tests, where we almost surely don't want to take on the challenge to format invalid code. We'll have to think about what to do with indentation of invalid code though. Do we all of a sudden start indenting pipelines in staircase when the file is invalid?

Reply:
>  An indenter should ideally work no matter the validity of the code

But an indenter requires you to first parse the code, and if the code doesn't parse, you are going to have trouble indenting

Reply:
I think that an indenter and a formatter are one and the same in my head, but it does not seem like you agree

Like to me people aren't going to be applying reindentation without also reformatting. At least, I don't think our tools are going to be allowing that. ruff doesn't, and it doesn't seem like the LSP has a request specifically for indentation, just formatting.

Reply:
> and if the code doesn't parse, you are going to have trouble indenting

That's why we use error-recovering parsers to power indenters (or regular expressions).

> I think that an indenter and a formatter are one and the same in my head, but it does not seem like you agree

They _can_ be, but a formatter is in general something more static than an indenter which is more dynamic. An indenter has to work no matter the state of the text because the user might be in the middle of writing code.

The two snippets you changed here are a good illustration, here is a common case the indenter needs to support:

```r
foo(bar +<cursor>)

# <enter>
foo(bar +
  <cursor>)
```

And that's a case where parentheses were auto-closed. The indenter should also know what to do with missing parentheses.

> it doesn't seem like the LSP has a request specifically for indentation, just formatting

In practice it's format-as-you-type but yes this is missing from the LSP, in part because editors treat indentation as latency-critical to provide a good experience to end users.

That said, fixing up indentation via the LSP works well enough in practice (as done in the Python extension with Pylance or positron-r with Ark).

We could split the task in two separate engines, an indenter that works on broken code and a formatter that requires valid code. In that case we would need to make sure the behaviours don't differ too much to avoid making it look like the editor is working against itself. There's also the question of the amount of work, though I think we have determined that the rule-based indenter does not require too much work.

The ideal case though is that the formatter itself can be used for indenting, so we need to think about whether it's really too expensive to let the formatter handle error-recovered code.

Reply:
Another use case to think about:

Copying parts of some code, and pasting it with formatting enabled. It would be nice to at least indent the code.

Reply:
> We could split the task in two separate engines, an indenter that works on broken code and a formatter that requires valid code.

Honestly I think that this doesn't sound too bad. But I'd be in favor of:
- A quite stupid indenter for Format on Type that just uses some basic rules and gets it right "most" of the time (like after a pipe or after a `+` like above). I guess it's ok that it uses the tree-sitter parser even on broken code. We will have to see. We've obviously determined that _just_ regex on the frontend isn't enough for this, so we probably do need the parser.
- A good formatter that only works on valid code, hooked up to Format on Save. In the 5% of the time that the indenter falls over, you just finish out the code to make it valid and hit "save" and then it gets fixed up and you keep going.

I think this is mostly how writing Rust feels and it doesn't bother me